modules/generator/processor/registry.go

```diff
@@ -0,0 +1,153 @@
+package processor
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	prometheus_model "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prometheus/model/exemplar"
+	prometheus_labels "github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Registry is a prometheus.Registerer that can gather metrics and push them directly into a
+// Prometheus storage.Appender.
+//
+// Currently, only counters and histograms are supported. Descriptions are ignored.
+type Registry struct {
+	prometheus.Registerer
+
+	gatherer prometheus.Gatherer
+
+	now func() time.Time
+}
+
+func NewRegistry(externalLabels map[string]string) Registry {
+	registry := prometheus.NewRegistry()
+
+	registerer := prometheus.WrapRegistererWith(externalLabels, registry)
+
+	return Registry{
+		Registerer: registerer,
+		gatherer:   registry,
+		now:        time.Now,
+	}
+}
+
+func (r *Registry) Gather(appender storage.Appender) error {
+	metricFamilies, err := r.gatherer.Gather()
+	if err != nil {
+		return err
+	}
+
+	timestamp := r.now().UnixMilli()
+
+	for _, metricFamily := range metricFamilies {
+
+		switch metricFamily.GetType() {
+		case prometheus_model.MetricType_COUNTER:
+			for _, metric := range metricFamily.GetMetric() {
+				labels := labelPairsToLabels(metric.Label)
+				labels = withLabel(labels, "__name__", metricFamily.GetName())
+
+				_, err := appender.Append(0, labels, timestamp, metric.GetCounter().GetValue())
+				if err != nil {
+					return err
+				}
+			}
+
+		case prometheus_model.MetricType_HISTOGRAM:
+			for _, metric := range metricFamily.GetMetric() {
+				labels := labelPairsToLabels(metric.Label)
+
+				histogram := metric.GetHistogram()
+
+				// _count
+				countLabels := withLabel(labels, "__name__", fmt.Sprintf("%s_count", metricFamily.GetName()))
+				_, err := appender.Append(0, countLabels, timestamp, float64(histogram.GetSampleCount()))
+				if err != nil {
+					return err
+				}
+
+				// _sum
+				sumLabels := withLabel(labels, "__name__", fmt.Sprintf("%s_sum", metricFamily.GetName()))
+				_, err = appender.Append(0, sumLabels, timestamp, histogram.GetSampleSum())
+				if err != nil {
+					return err
+				}
+
+				addedInfBucket := false
+
+				// _bucket
+				bucketLabels := withLabel(labels, "__name__", fmt.Sprintf("%s_bucket", metricFamily.GetName()))
+				for _, bucket := range histogram.GetBucket() {
+
+					if bucket.GetUpperBound() == math.Inf(1) {
+						addedInfBucket = true
+					}
+
+					// TODO make a complete copy of this slice, not sure why this is needed
+					//  it works without in registry_test.go but fails in servicegraphs_test.go 🙃
+					//  in the servicegraphs test the le labels gets overwritten with the latest value 🤷
```

Comment:
Yeah, it is. I can take a deeper look at this again. I think the issue is that we take a shallow copy of the slice and append a new label to each copy. Since the shallow copies share the same underlying buffer, the appends overwrite each other.

Reply:
Okay, tracked down the issue. The other tests fail when they add 2 additional labels, the registry tests only adds at most 1. If you add 2 labels the capacity of the underlying slice grows and appending the `le` label does not allocate a new buffer -> memory is shared between different buckets and the value of the `le` label gets overwritten.
Summary: this is a bug caused by poor slice handling and it works sometimes because of how the slice is reallocated when it grows 😬

Reply:
Fixed this now 🙂